feat: Support private accounts on `MockChain`

igamigo opened the pull request:
Closes #1314

igamigo commented:
@PhilippGackstatter I think this is all that's really needed to support private accounts on `MockChain`. 

The `committed_account` mapping is now a bit opaque though: Since it holds the latest known state, it doesn't really get updated on every committed transaction for any private account. It is still useful to add private accounts to it to hold the authenticator, but private accounts should only be executed against through `TxContextInput::Account` in order to not execute with a stale state. I updated some docs, and could also update the `build_tx_context()` function to disallow creating contexts for private accounts with `TxContextInput::AccountId` if you think that's a good idea (as a consequence of this, a bunch of tests need to be updated to use public accounts).

PhilippGackstatter reviewed (approved):
Looks good to me!

I think this works, but I wonder if `MockChain::build_tx_context` should panic if `TxContextInput::AccountId` is passed with a private account ID. E.g. `assert!(
                    !account_id.is_private(),
                    "private account ID should not be passed to build_tx_context, pass the full account instead"
                );`. Long-term this should return an Error instead of panicking, but that is part of https://github.com/0xMiden/miden-base/issues/980. Right now, this makes 7 tests fail, but it would probably be a good measure to avoid executing with stale state going forward.

Additionally, and somewhat alternatively, we could also not store the full `MockAccount` for private accounts in the chain and instead store only a mapping of 
```rust
/// AccountId |-> (seed, authenticator) mapping.
account_helpers: BTreeMap<AccountId, (Option<Word>, Option<BasicAuthenticator<ChaCha20Rng>>)>,
```
as shown in the diff in https://github.com/0xMiden/miden-base/issues/980#issuecomment-2869531226. This would be changed for all accounts, not just private ones.

I think that's the better solution long-term, so we don't store private account state at all and so there is no risk to accidentally fetch stale state. That should be a separate PR though because it would be a larger refactor.

igamigo commented:
Yeah, agreed. I had added something like that on [624790f](https://github.com/0xMiden/miden-base/pull/1453/commits/624790fdcb2ddb610e25bc9b2cda584b27948416) but deferred it to #980, though it makes more sense to do it here so I re-added it and fixed the tests.
